shared/ssz/decode.go

```diff
@@ -277,26 +277,27 @@ func makeStructDecoder(typ reflect.Type) (decoder, error) {
 	return decoder, nil
 }
 
-// Notice: Currently we don't support nil pointer:
-// - Input for encoding must not contain nil pointer
-// - Output for decoding will never contain nil pointer
-// (Not to be confused with empty slice. Empty slice is supported)
 func makePtrDecoder(typ reflect.Type) (decoder, error) {
 	elemType := typ.Elem()
 	elemSSZUtils, err := cachedSSZUtilsNoAcquireLock(elemType)
 	if err != nil {
 		return nil, err
 	}
+
+	// TODO(1461): The encoding of nil pointer isn't defined in the spec.
+	// After considered the use case in Prysm, we've decided that:
+	// - We assume we will only encode/decode pointer of array, slice or struct.
+	// - The encoding for nil pointer shall be 0x00000000.
+
 	decoder := func(r io.Reader, val reflect.Value) (uint32, error) {
-		newVal := val
-		if val.IsNil() {
-			newVal = reflect.New(elemType)
-		}
+		newVal := reflect.New(elemType)
 		elemDecodeSize, err := elemSSZUtils.decoder(r, newVal.Elem())
 		if err != nil {
 			return 0, fmt.Errorf("failed to decode to object pointed by pointer: %v", err)
 		}
-		val.Set(newVal)
+		if elemDecodeSize != lengthBytes {
```

Comment:
shouldnt this be `==` ?

Reply:
+1 here @houjieth

Reply:
Yeah it's a little tricky. I mentioned that I will assume we will only support pointer that points to slice/array/struct. If the pointer is nil, the encoding should be `0x00000000`, which is  `lengthBytes` bytes. If the pointer is not nil, the encoding should be the encoding of the object the pointer points to, which should contain `lengthBytes` bytes header followed by a number of bytes. So the encoding length will be greater than `lengthBytes`.

So `elemDecodeSize == lengthBytes` means we were encoding nil pointer, `elemDecodeSize > lengthBytes` means we were encoding a non-nil pointer.

Reply:
I will change it to `elemDecodeSize > lengthBytes` so it looks less confusing.

Reply:
Change submitted.